vsintegration/src/FSharp.Editor/TaskList/TaskListService.fs

```diff
@@ -0,0 +1,73 @@
+﻿
+// Microsoft.CodeAnalysis.ExternalAccess.FSharp.TaskList.FSharpTaskListService
+
+namespace Microsoft.VisualStudio.FSharp.Editor
+
+open System
+open System.ComponentModel.Composition
+open Microsoft.CodeAnalysis.Text
+open FSharp.Compiler.CodeAnalysis
+open System.Runtime.InteropServices
+open Microsoft.CodeAnalysis.ExternalAccess.FSharp.Editor
+open Microsoft.CodeAnalysis.ExternalAccess.FSharp.TaskList
+open FSharp.Compiler
+open System.Collections.Immutable
+
+open System.Diagnostics
+
+[<Export(typeof<IFSharpTaskListService>)>]
+type internal FSharpTaskListService 
+    [<ImportingConstructor>]
+    (
+    ) as this=
+
+    let getDefines (doc:Microsoft.CodeAnalysis.Document) = 
+        asyncMaybe { 
+            let! _, _, parsingOptions, _ = doc.GetFSharpCompilationOptionsAsync(nameof(FSharpTaskListService)) |> liftAsync 
+            return CompilerEnvironment.GetConditionalDefinesForEditing parsingOptions }
+        |> Async.map (Option.defaultValue [])
+
+    member _.GetTaskListItems(doc:Microsoft.CodeAnalysis.Document, descriptors: (string*'T)[], ct) = 
+            backgroundTask{
+                let foundTaskItems = ImmutableArray.CreateBuilder(initialCapacity=0)
+                let! sourceText = doc.GetTextAsync(ct)             
+                let! defines = doc |> getDefines
+               
+                for line in sourceText.Lines do  
+                 
+                    let unfilteredTokens = Tokenizer.tokenizeLine(doc.Id, sourceText, line.Span.Start, doc.FilePath, defines) 
+                    let granularTokens = unfilteredTokens |> Array.filter(fun t -> t.Kind = LexerSymbolKind.Comment)               
+
+                    let contractedTokens = 
+                        ([],granularTokens) 
+                        ||> Array.fold (fun acc token -> 
+                            let token = {|Left = token.LeftColumn; Right = token.RightColumn|} 
+                            match acc with
+                            | [] -> [token]
+                            | head :: tail when token.Left-head.Right <= 1   -> {|token with Left = head.Left|} :: tail
+                            | _  -> token :: acc )
+
+                    for ct in contractedTokens do    
```

Comment:
`ct` hides cancellation token. I wonder if this service would benefit from cancellation checks inside the loop? Probably depends on how often it runs and how fast the tokenizer is.

Reply:
The implementation within Tokenizer did not have cancellation support before.
At the same time, it had bug in cache-miss scenario (when data for line X is not up to date, in only re-scanned line X. However, it must rescan from last valid line, since lexstate can flow between lines in situations like multiline strings, ifdefs etc.) and after fixing it, I think cancellation support now makes sense.

Since it can potentially re-scan 0..n lines in a cache-miss scenario (after the bugfix), adding cancellation check into that loop feels natural.

==> I will make an update.

Reply:
It ended up being a refactoring affecting more users of Tokenizer, I would now prefer to do the overall tokenization cancellation as a separate PR instead of packing it up here.

Reply:
I will put it in here, will split in two if people complain a lot.

Reply:
And thanks for calling this out, @majocha  :)